src/sentry/types/region.py

```diff
@@ -1,8 +1,120 @@
+from __future__ import annotations
+
 from dataclasses import dataclass
+from enum import Enum
+from typing import TYPE_CHECKING, Iterable
+
+from sentry.silo import SiloMode
+
+if TYPE_CHECKING:
+    from sentry.models import Organization
+
+
+class RegionCategory(Enum):
+    GEOGRAPHIC = "GEOGRAPHIC"
+    SINGLE_TENANT = "SINGLE_TENANT"
 
 
-@dataclass
+@dataclass(frozen=True, eq=True)
 class Region:
+    """A region of the Sentry platform, hosted by a region silo."""
+
     name: str
-    subdomain: str
-    is_private: bool = False
+    """The region's unique identifier."""
+
+    id: int
+    """The region's unique numeric representation.
+
+    This number is used for composing "snowflake" IDs, and must fit inside the
+    maximum bit length specified by our snowflake ID schema.
+    """
+
+    address: str
+    """The address of the region's silo.
+
+    In a production environment, this would typically be a subdomain (e.g.,
+    "eu.sentry.io"). In a development environment, it may be useful to define
+    region addresses as a custom hostname or port number (e.g,
+    "localhost:8001").
+    """
+
+    category: RegionCategory
+    """The region's category."""
+
+    def __post_init__(self) -> None:
+        from sentry.utils.snowflake import REGION_ID
+
+        REGION_ID.validate(self.id)
+
+    def to_url(self, path: str) -> str:
+        return self.address + path
```

Comment:
As above, this is just conceptual. We'll handle "do we need to join them with a slash" and other annoying details when the time comes.

Reply:
On the other hand, this definitely seems more related specifically to endpoint API and not necessary the Region as a whole.  I imagine there being a "SiloHTTPApiClient" that handles that and can be constructed from a Region, but not assuming there is a singular such that associated with a Region.